File: ukulele_chords.js
// ukulele_chords.js — JavaScript версия

class UkuleleChord {
    constructor(tuning = 'GCEA') {
        this.tuning = tuning;
        this.strings = tuning === 'GCEA' ? ['G', 'C', 'E', 'A'] :
                       tuning === 'ADF#B' ? ['A', 'D', 'F#', 'B'] :
                       ['D', 'G', 'B', 'E'];
        this.chords = this._initChords();
    }

    _initChords() {
        return {
            C: { major: {3:0, 2:0, 1:0, 0:3}, minor: {3:0, 2:3, 1:3, 0:3}, seventh: {3:0, 2:0, 1:0, 0:1} },
            D: { major: {3:2, 2:2, 1:2, 0:0}, minor: {3:2, 2:1, 1:2, 0:0}, seventh: {3:2, 2:1, 1:2, 0:0} },
            E: { major: {3:1, 2:4, 1:0, 0:4}, minor: {3:0, 2:4, 1:3, 0:2}, seventh: {3:0, 2:0, 1:0, 0:4} },
            F: { major: {3:2, 2:0, 1:1, 0:0}, minor: {3:1, 2:0, 1:1, 0:3}, seventh: {3:2, 2:0, 1:1, 0:3} },
            G: { major: {3:0, 2:2, 1:3, 0:2}, minor: {3:0, 2:2, 1:3, 0:3}, seventh: {3:0, 2:2, 1:3, 0:2} },
            A: { major: {3:0, 2:1, 1:2, 0:0}, minor: {3:0, 2:0, 1:2, 0:2}, seventh: {3:0, 2:0, 1:2, 0:1} },
            B: { major: {3:4, 2:3, 1:2, 0:0}, minor: {3:4, 2:2, 1:2, 0:2}, seventh: {3:4, 2:1, 1:2, 0:2} }
        };
    }

    getChord(root, type) {
        if (this.chords[root] && this.chords[root][type]) {
            return this.chords[root][type];
        }
        return null;
    }

    printFretboard(pos) {
        console.log(`\n   ${this.strings.join(' ')}`);
        console.log(`   ${'─'.repeat(this.strings.length * 2 + 1)}`);
        for (let fret = 0; fret < 5; fret++) {
            let line = fret === 0 ? '  ' : `${fret} `;
            line += ' |';
            for (let str = 0; str < 4; str++) {
                if (pos.hasOwnProperty(str) && pos[str] === fret) {
                    line += ' ● |';
                } else if (pos.hasOwnProperty(str) && pos[str] < fret) {
                    line += '   |';
                } else {
                    line += '   |';
                }
            }
            console.log(line);
        }
        console.log('\nАппликатура:');
        for (let str = 0; str < 4; str++) {
            if (pos.hasOwnProperty(str)) {
                if (pos[str] === 0) {
                    console.log(`  ${this.strings[str]}: открытая`);
                } else {
                    console.log(`  ${this.strings[str]}: ${pos[str]}-й лад`);
                }
            }
        }
    }

    listChords() {
        console.log('Доступные аккорды для укулеле:');
        for (const root of Object.keys(this.chords).sort()) {
            const types = Object.keys(this.chords[root]);
            console.log(`  ${root}: ${types.join(', ')}`);
        }
    }
}

function main() {
    const args = process.argv.slice(2);
    let chord = 'C', type = 'major', tuning = 'GCEA', list = false;

    for (let i = 0; i < args.length; i++) {
        if (args[i] === '--chord') chord = args[++i];
        else if (args[i] === '--type') type = args[++i];
        else if (args[i] === '--tuning') tuning = args[++i];
        else if (args[i] === '--list') list = true;
    }

    const generator = new UkuleleChord(tuning);

    if (list) {
        generator.listChords();
        return;
    }

    const pos = generator.getChord(chord, type);
    if (!pos) {
        console.log(`❌ Аккорд ${chord} ${type} не найден.`);
        console.log('Используйте --list для просмотра всех доступных аккордов.');
        return;
    }

    console.log(`\n🎸 Аккорд: ${chord} (${type}) | Настройка: ${tuning}`);
    generator.printFretboard(pos);
}

if (require.main === module) main();

File: ukulele_chords.rs
// ukulele_chords.rs — Rust версия

use std::collections::HashMap;

struct UkuleleChord {
    tuning: String,
    strings: Vec<String>,
    chords: HashMap<String, HashMap<String, HashMap<usize, usize>>>,
}

impl UkuleleChord {
    fn new(tuning: &str) -> Self {
        let strings = match tuning {
            "GCEA" => vec!["G".to_string(), "C".to_string(), "E".to_string(), "A".to_string()],
            "ADF#B" => vec!["A".to_string(), "D".to_string(), "F#".to_string(), "B".to_string()],
            _ => vec!["D".to_string(), "G".to_string(), "B".to_string(), "E".to_string()],
        };
        let mut chord = UkuleleChord {
            tuning: tuning.to_string(),
            strings,
            chords: HashMap::new(),
        };
        chord.init_chords();
        chord
    }

    fn init_chords(&mut self) {
        let mut c = HashMap::new();
        c.insert("major".to_string(), {
            let mut m = HashMap::new();
            m.insert(3, 0); m.insert(2, 0); m.insert(1, 0); m.insert(0, 3);
            m
        });
        c.insert("minor".to_string(), {
            let mut m = HashMap::new();
            m.insert(3, 0); m.insert(2, 3); m.insert(1, 3); m.insert(0, 3);
            m
        });
        c.insert("seventh".to_string(), {
            let mut m = HashMap::new();
            m.insert(3, 0); m.insert(2, 0); m.insert(1, 0); m.insert(0, 1);
            m
        });
        self.chords.insert("C".to_string(), c);

        let mut d = HashMap::new();
        d.insert("major".to_string(), {
            let mut m = HashMap::new();
            m.insert(3, 2); m.insert(2, 2); m.insert(1, 2); m.insert(0, 0);
            m
        });
        d.insert("minor".to_string(), {
            let mut m = HashMap::new();
            m.insert(3, 2); m.insert(2, 1); m.insert(1, 2); m.insert(0, 0);
            m
        });
        d.insert("seventh".to_string(), {
            let mut m = HashMap::new();
            m.insert(3, 2); m.insert(2, 1); m.insert(1, 2); m.insert(0, 0);
            m
        });
        self.chords.insert("D".to_string(), d);

        let mut e = HashMap::new();
        e.insert("major".to_string(), {
            let mut m = HashMap::new();
            m.insert(3, 1); m.insert(2, 4); m.insert(1, 0); m.insert(0, 4);
            m
        });
        e.insert("minor".to_string(), {
            let mut m = HashMap::new();
            m.insert(3, 0); m.insert(2, 4); m.insert(1, 3); m.insert(0, 2);
            m
        });
        e.insert("seventh".to_string(), {
            let mut m = HashMap::new();
            m.insert(3, 0); m.insert(2, 0); m.insert(1, 0); m.insert(0, 4);
            m
        });
        self.chords.insert("E".to_string(), e);

        let mut f = HashMap::new();
        f.insert("major".to_string(), {
            let mut m = HashMap::new();
            m.insert(3, 2); m.insert(2, 0); m.insert(1, 1); m.insert(0, 0);
            m
        });
        f.insert("minor".to_string(), {
            let mut m = HashMap::new();
            m.insert(3, 1); m.insert(2, 0); m.insert(1, 1); m.insert(0, 3);
            m
        });
        f.insert("seventh".to_string(), {
            let mut m = HashMap::new();
            m.insert(3, 2); m.insert(2, 0); m.insert(1, 1); m.insert(0, 3);
            m
        });
        self.chords.insert("F".to_string(), f);

        let mut g = HashMap::new();
        g.insert("major".to_string(), {
            let mut m = HashMap::new();
            m.insert(3, 0); m.insert(2, 2); m.insert(1, 3); m.insert(0, 2);
            m
        });
        g.insert("minor".to_string(), {
            let mut m = HashMap::new();
            m.insert(3, 0); m.insert(2, 2); m.insert(1, 3); m.insert(0, 3);
            m
        });
        g.insert("seventh".to_string(), {
            let mut m = HashMap::new();
            m.insert(3, 0); m.insert(2, 2); m.insert(1, 3); m.insert(0, 2);
            m
        });
        self.chords.insert("G".to_string(), g);

        let mut a = HashMap::new();
        a.insert("major".to_string(), {
            let mut m = HashMap::new();
            m.insert(3, 0); m.insert(2, 1); m.insert(1, 2); m.insert(0, 0);
            m
        });
        a.insert("minor".to_string(), {
            let mut m = HashMap::new();
            m.insert(3, 0); m.insert(2, 0); m.insert(1, 2); m.insert(0, 2);
            m
        });
        a.insert("seventh".to_string(), {
            let mut m = HashMap::new();
            m.insert(3, 0); m.insert(2, 0); m.insert(1, 2); m.insert(0, 1);
            m
        });
        self.chords.insert("A".to_string(), a);

        let mut b = HashMap::new();
        b.insert("major".to_string(), {
            let mut m = HashMap::new();
            m.insert(3, 4); m.insert(2, 3); m.insert(1, 2); m.insert(0, 0);
            m
        });
        b.insert("minor".to_string(), {
            let mut m = HashMap::new();
            m.insert(3, 4); m.insert(2, 2); m.insert(1, 2); m.insert(0, 2);
            m
        });
        b.insert("seventh".to_string(), {
            let mut m = HashMap::new();
            m.insert(3, 4); m.insert(2, 1); m.insert(1, 2); m.insert(0, 2);
            m
        });
        self.chords.insert("B".to_string(), b);
    }

    fn get_chord(&self, root: &str, chord_type: &str) -> Option<HashMap<usize, usize>> {
        if let Some(root_map) = self.chords.get(root) {
            if let Some(pos) = root_map.get(chord_type) {
                return Some(pos.clone());
            }
        }
        None
    }

    fn print_fretboard(&self, pos: &HashMap<usize, usize>) {
        println!("\n   {}", self.strings.join(" "));
        println!("   {}", "─".repeat(self.strings.len() * 2 + 1));
        for fret in 0..5 {
            let mut line = if fret == 0 { "  ".to_string() } else { format!("{} ", fret) };
            line.push_str(" |");
            for str in 0..4 {
                if let Some(&val) = pos.get(&str) {
                    if val == fret {
                        line.push_str(" ● |");
                    } else if val < fret {
                        line.push_str("   |");
                    } else {
                        line.push_str("   |");
                    }
                } else {
                    line.push_str("   |");
                }
            }
            println!("{}", line);
        }
        println!("\nАппликатура:");
        for str in 0..4 {
            if let Some(&val) = pos.get(&str) {
                if val == 0 {
                    println!("  {}: открытая", self.strings[str]);
                } else {
                    println!("  {}: {}-й лад", self.strings[str], val);
                }
            }
        }
    }

    fn list_chords(&self) {
        println!("Доступные аккорды для укулеле:");
        let mut roots: Vec<_> = self.chords.keys().collect();
        roots.sort();
        for root in roots {
            let types: Vec<_> = self.chords[root].keys().collect();
            println!("  {}: {}", root, types.join(", "));
        }
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let mut chord = "C".to_string();
    let mut ctype = "major".to_string();
    let mut tuning = "GCEA".to_string();
    let mut list = false;

    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--chord" => { chord = args[i+1].clone(); i += 2; }
            "--type" => { ctype = args[i+1].clone(); i += 2; }
            "--tuning" => { tuning = args[i+1].clone(); i += 2; }
            "--list" => { list = true; i += 1; }
            _ => { i += 1; }
        }
    }

    let generator = UkuleleChord::new(&tuning);

    if list {
        generator.list_chords();
        return;
    }

    if let Some(pos) = generator.get_chord(&chord, &ctype) {
        println!("\n🎸 Аккорд: {} ({}) | Настройка: {}", chord, ctype, tuning);
        generator.print_fretboard(&pos);
    } else {
        println!("❌ Аккорд {} {} не найден.", chord, ctype);
        println!("Используйте --list для просмотра всех доступных аккордов.");
    }
}
